agent/qrexec-agent.c

```diff
@@ -119,6 +121,43 @@ static struct pam_conv conv = {
     NULL
 };
 #endif
+
+static void set_cloexec_on_all_fds(void) {
+    int const procfd = open("/proc/self/fd", O_DIRECTORY|O_RDONLY|O_CLOEXEC|O_NOCTTY);
+    if (procfd == -1) {
+        PERROR("open /proc/self/fd");
+        exit(1);
+    }
+    DIR *dir = fdopendir(procfd);
+    if (!dir) {
+        PERROR("fdopendir");
+        exit(1);
+    }
+    for (;;) {
+        errno = 0;
+        struct dirent *entry = readdir(dir);
+        char *endptr;
+        long fd_num;
+        if (!entry) {
+            if (errno)
+                err(1, "readdir(/proc/self/fd)");
+            break;
+        }
+        errno = 0;
+        fd_num = strtol(entry->d_name, &endptr, 10);
+        if (errno || !endptr || *endptr || fd_num < 0 || fd_num > INT_MAX)
+            errx(1, "bad number in /proc/self/fd");
```

Comment:
This always trips, I'm pretty sure it's about `.` or `..`.

Anyway, setting all FDs to `O_CLOEXEC` sounds fragile (there may be cases when inheriting open FD would be desirable, even if currently it isn't used), and also could hide other issues. I'd prefer to do it properly:
 - inspect what FDs are currently passed on to children, and for each consciously decide whether `O_CLOEXEC` is desirable (or maybe it shouldn't be left open at all?)
 - (optionally) add a test that checks for FD leaks - we do have proper unit tests in this repo, so it shouldn't be too hard. see `qrexec/tests/socket/agent.py` for example